File: Innerproc2.0-CXX/message.cpp
#include "message.h"
#include <jsoncpp/json/json.h>

/**
 * parse()
 * @param data
 * @param size
 * @return
 *
 * {
 * 	 name: line_busy,
 * 	 id: 322213123,
 *   values:{
 *		url: xxx,
 *		type: 10,
 *		...
 *   }
 * }
 */
 
typedef std::function< std::shared_ptr<Message> (const Json::Value& root) > ParseFunc;

std::vector<ParseFunc> parse_func_list={
	MessageJoinFamily::parse,
	MessageJoinReject::parse,
	MessageJoinAccept::parse,
	MessageCall::parse,
	MessageCallIn::parse,
	MessageCallOut::parse,
	MessageCallAccept::parse,
	MessageCallReject::parse,
	MessageCallEnd::parse,
	MessageCallKeep::parse,
//	MessageLineBusy::parse,
	MessageHeartbeat::parse,
	MessageEmergency::parse,
};

Message::Ptr MessageJsonParser::parse(const char * data,size_t size){
	Json::Reader reader;
	Json::Value root;
	Message::Ptr msg;
	if (reader.parse(data, root)){
		for(auto func:parse_func_list){
			msg = func(root);
			if(msg){
				msg->id_ = root["id"].asString();
				
				
				break;
			}
		}
	}
	return msg;
}


std::shared_ptr<Message> MessageJoinFamily::parse(const Json::Value& root){
	std::shared_ptr<Message> result;
	if( root["name"].asString() ==  "join_family"){
		std::shared_ptr<MessageJoinFamily> msg = std::make_shared<MessageJoinFamily>();
		Json::Value values = root["values"];
		msg->token = values["token"].asString();
		msg->id =  values["id"].asString();
		msg->type = values["type"].asString();
		result = msg ;
	}
	return result;
}

std::shared_ptr<Message> MessageJoinReject::parse(const Json::Value& root){
	return std::shared_ptr<MessageJoinReject>();
}

std::shared_ptr<Message> MessageHeartbeat::parse(const Json::Value& root){
	std::shared_ptr<Message> msg;
	if( root["name"].asString() == "heartbeat"){
		msg = std::make_shared<MessageHeartbeat>();
	}
	return msg;
}

std::shared_ptr<Message> MessageCallReject::parse(const Json::Value& root){
	std::shared_ptr<Message> result;
	if( root["name"].asString() == "call_reject"){
		auto msg = std::make_shared<MessageCallReject>();
		Json::Value values = root["values"];
		msg->code =  (MessageCallReject::Reason)values.get("code",(int)MessageCallReject::Reason::undefined).asInt();
		msg->message = values.get("message","").asString();
		result = msg;
	}
	return result;
}



std::shared_ptr<Message> MessageAlarm::parse(const Json::Value& root){
	
	return std::shared_ptr<Message>();
}


std::shared_ptr<Message> MessageEmergency::parse(const Json::Value& root) {
	
	return std::shared_ptr<Message>();
}


Json::Value MessageCallIn::values(){
	Json::Value value;
	value["sid"] = "";
	Json::Value src,dest;
	src["id"] =  this->src.id;
	src["type"] = this->src.type;
	src["ip"] = this->src.ip;
	src["port"] = this->src.port;
	src["stream_audio_url"] = this->src.stream_audio_url;
	src["stream_video_url"] = this->src.stream_video_url;
	
	dest["id"] =  this->dest.id;
	dest["type"] = this->dest.type;
	dest["ip"] = this->dest.ip;
	dest["port"] = this->dest.port;
	dest["stream_audio_url"] = this->dest.stream_audio_url;
	dest["stream_video_url"] = this->dest.stream_video_url;
	
	value["src"] = src;
	value["dest"] = dest;
	
	return value;
}

/*
 *
 *
 * http://open-source-parsers.github.io/jsoncpp-docs/doxygen/index.html
 *
 */

File: Tests/message.py
#coding: utf-8

import json
from mantis.fundamental.utils.useful import hash_object,object_assign
from mantis.fundamental.network.message import JsonMessage as Message
"""
{
    id,
    name,
    values:{
        ..
    }:
}

"""

"""
enum CallPeerType{
	INNER_BOX = 1,
	INNER_SCREEN = 2,
	OUTER_BOX = 3,
	PROPERTY_APP = 4,
	PROPERTY_CENTER = 5
};

"""

class CallPeerType(object):

    Unknown = 0
    INNER_BOX = 1
    INNER_SCREEN = 2
    OUTER_BOX = 3
    PROPERTY_APP = 4
    PROPERTY_CENTER = 5

class CallEndpoint(object):
    def __init__(self):
        self.ip = ''
        self.port = 0
        # self.room_id =''
        self.id = ''
        self.type = CallPeerType.Unknown
        self.video_stream_id = ''
        self.audio_stream_id = ''



class MessageJoinFamily(Message):
    """加入家庭"""
    NAME = 'join_family'
    def __init__(self):
        Message.__init__(self,self.NAME)
        self.token = ''
        self.id = ''
        self.type = ''

class MessageJoinReject(Message):
    """加入家庭拒绝"""
    NAME = 'join_reject'
    def __init__(self):
        Message.__init__(self,self.NAME)
        self.reason = ''

class MessageJoinAccept(Message):
    """加入家庭成功"""
    NAME ='join_accept'
    def __init__(self):
        Message.__init__(self,self.NAME)
        self.session_id = ''    # 会话id
        self.room_id = ''       # 房号

class MessageCall(Message):
    """呼叫"""
    NAME = 'call_req'
    def __init__(self):
        Message.__init__(self,self.NAME)
        self.sid = ''
        self.src = CallEndpoint()
        self.dest = CallEndpoint()

    def assign(self,data):
        object_assign(self.src,data.get('src',{}))
        object_assign(self.dest,data.get('dest',{}))

    def values(self):
        src = hash_object(self.src)
        dest = hash_object(self.dest)
        return dict(sid=self.sid,src=src,dest=dest)

# class MessageCallIn(MessageCall):
#     """呼叫进入"""
#     NAME = 'call_in'
#     def __init__(self):
#         MessageCall.__init__(self,self.NAME)
#
#
# class MessageCallOut(MessageCall):
#     """呼叫外出"""
#     NAME = 'call_out'
#     def __init__(self):
#         MessageCall.__init__(self,self.NAME)

class MessageCallAccept(Message):
    """呼叫接听"""
    NAME = 'call_accept'
    def __init__(self):
        Message.__init__(self,self.NAME)

class MessageCallReject(Message):
    """呼叫拒绝"""
    NAME = 'call_reject'
    def __init__(self):
        Message.__init__(self,self.NAME)

class MessageCallEnd(Message):
    """呼叫结束"""
    NAME = 'call_end'
    def __init__(self):
        Message.__init__(self,self.NAME)

class MessageCallKeep(Message):
    """呼叫状态保持"""
    NAME = 'call_keep'
    def __init__(self):
        Message.__init__(self,self.NAME)

class MessageHeartbeat(Message):
    """呼叫进入"""
    NAME = 'heartbeat'
    def __init__(self):
        Message.__init__(self,self.NAME)


class MessageEmergency(Message):
    """呼叫进入"""
    NAME = 'emergency'
    def __init__(self):
        Message.__init__(self,self.NAME)

class MessageOpenDoor(Message):
    """开门"""
    NAME = 'open_door'
    def __init__(self):
        Message.__init__(self,self.NAME)

MessageClsDict ={}

def registerMessageObject(msgcls):
    MessageClsDict[msgcls.NAME] = msgcls


for key,value in locals().items():
    if key.find('Message')==0 and key not in ('MessageClsDict','Message','MessageType','MessageSplitter'):
        registerMessageObject(value)

def parseMessage(data):
    print data
    if  isinstance(data,str):
        data = json.loads(data)

    message = data.get('name')
    msgcls = MessageClsDict.get(message)
    if not msgcls:
        print 'Message Type unKnown. value:{}'.format(message)
        return None
    data = data.get('values',{})
    msg = msgcls()
    msg.assign(data)
    return msg


if __name__ == '__main__':
    print MessageClsDict

